src/loader.js

```diff
@@ -0,0 +1,91 @@
+let assume = require('assume');
+let debug = require('debug')('taskcluster:loader');
+let Promise = require('promise');
+
+function loader(componentDirectory, requiredVirtualComponents) {
+  /**
+   * If present, this is a list of component names that a loader
+   * must define.  These are likely things like configuration values
+   */
+  if (requiredVirtualComponents) {
+    assume(requiredVirtualComponents).is.array();
+  } else {
+    requiredVirtualComponents = [];
+  }
+
+  return function(virtualComponents) {
+    let initializedComponents = {};
+
+    function loadComponent(name, visited) {
+      assume(name).to.be.ok();
+
+      if (!visited) {
```

Comment:
didn't test it but I think the concept works... as we don't have cycles this should be safe...